File: backend/app/models/memory.py
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


class MemoryField(BaseModel):
    value: Optional[str] = None
    confidence: int = Field(default=0, ge=0, le=100)
    sources: List[str] = Field(default_factory=list)


class TripMemoryBase(BaseModel):
    destination: Optional[MemoryField] = None
    dates: Optional[MemoryField] = None
    budget: Optional[MemoryField] = None
    pace: Optional[MemoryField] = None
    duration: Optional[MemoryField] = None


class TripMemoryUpdate(TripMemoryBase):
    pass


class TripMemoryInDB(TripMemoryBase):
    id: Optional[str] = Field(default=None, alias="_id")
    trip_id: Optional[str] = Field(default=None, alias="tripId")
    updated_at: Optional[datetime] = None

    class Config:
        populate_by_name = True


class TripMemory(TripMemoryBase):
    id: str
    trip_id: str
    updated_at: datetime

    class Config:
        from_attributes = True

File: backend/src/models/memory.ts
import { z } from 'zod';

export const MemoryFieldSchema = z.object({
  value: z.string().optional(),
  confidence: z.number().min(0).max(100).default(0),
  sources: z.array(z.string()).default([]),
});

export const TripMemoryBaseSchema = z.object({
  destination: MemoryFieldSchema.optional(),
  dates: MemoryFieldSchema.optional(),
  budget: MemoryFieldSchema.optional(),
  pace: MemoryFieldSchema.optional(),
  duration: MemoryFieldSchema.optional(),
});

export const TripMemoryUpdateSchema = TripMemoryBaseSchema;

export type MemoryField = z.infer<typeof MemoryFieldSchema>;
export type TripMemoryBase = z.infer<typeof TripMemoryBaseSchema>;
export type TripMemoryUpdate = z.infer<typeof TripMemoryUpdateSchema>;

export interface TripMemoryInDB extends TripMemoryBase {
  _id?: string;
  tripId?: string;
  updatedAt?: Date;
}

export interface TripMemory extends TripMemoryBase {
  id: string | null;
  trip_id: string;
  updated_at: Date | null;
}
